libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx

```diff
@@ -0,0 +1,165 @@
+import {
+  Autocomplete,
+  AutocompleteProps,
+  CircularProgress,
+} from '@mui/material';
+import { useField, useFormikContext } from 'formik';
+import { useEffect, useState } from 'react';
+import tw from 'twin.macro';
+import { FormObjectValue } from '@unocha/hpc-data';
+import { StyledTextField } from './text-field';
+
+const StyledAutocomplete = tw(Autocomplete)`
+    min-w-[10rem]
+    w-full`;
+
+type AsyncAutocompleteSelectProps = {
+  name: string;
+  label: string;
+  placeholder?: string;
+  fnPromise: ({ query }: { query: string }) => Promise<Array<FormObjectValue>>;
+  isMulti?: boolean;
+  isAutocompleteAPI?: boolean;
+  error?: (metaError: string) => string | undefined;
+  required?: boolean;
+};
+const AsyncAutocompleteSelect = ({
+  name,
+  label,
+  placeholder,
+  fnPromise,
+  isMulti,
+  error,
+  isAutocompleteAPI,
+  required,
+  ...otherProps
+}: AsyncAutocompleteSelectProps) => {
+  const [open, setOpen] = useState(false);
+  const [inputValue, setInputValue] = useState('');
+  const { setFieldValue } = useFormikContext<Array<FormObjectValue>>();
+  const [field, meta] = useField<Array<FormObjectValue>>(name);
+  const [options, setOptions] = useState<Array<FormObjectValue>>([]);
+  const [data, setData] = useState<Array<FormObjectValue>>([]);
+  const [isFetch, setIsFetch] = useState(false);
+  const loading =
+    open && !isFetch && (!isAutocompleteAPI || inputValue.length >= 3);
+
+  useEffect(() => {
+    let active = true;
+    if (isAutocompleteAPI && (inputValue === '' || inputValue.length < 3)) {
+      setOptions([]);
+      setData([]);
+      setIsFetch(false);
+      return undefined;
+    }
+    if (data.length > 0 && (inputValue.length >= 3 || !isAutocompleteAPI)) {
+      setOptions(
+        data.filter((x) =>
+          x.displayLabel.toUpperCase().includes(inputValue.toUpperCase())
+        )
+      );
+    }
+
+    if (!loading) {
+      return undefined;
+    }
+    (async () => {
+      try {
+        const response = await fnPromise({
+          query: inputValue,
+        });
+        setData(response);
+        console.log(response);
+        if (active) {
+          setOptions(response);
+        }
+        setIsFetch(true);
+      } catch (error) {
+        console.error(error);
+      }
+    })();
+
+    return () => {
+      active = false;
+    };
+  }, [loading, inputValue]);
+
+  useEffect(() => {
+    if (!open && isAutocompleteAPI) {
+      setOptions([]);
+      setData([]);
+      setIsFetch(false);
+    }
+  }, [open]);
+
+  const configAutocomplete: AutocompleteProps<
+    FormObjectValue,
+    boolean,
+    boolean,
+    boolean
+  > = {
+    ...field,
+    ...otherProps,
+    multiple: isMulti,
+    placeholder: placeholder,
```

Comment:
We can use more modern syntax (here and in a few places below):
```suggestion
    placeholder,
```